### PR TITLE
Move to .NET 8 WASI support

### PR DESCRIPTION
**Pull request by RReverser**
# Description of Changes

.NET 8 shipped last week with built-in WASI support. It's still marked as experimental because some things - like better import/export handling that we could use - are still being worked on, but it's already better than using the `Wasi.Sdk` package which is now soft-deprecated and has a number of bugs that nobody's going to fix anymore.

This change is pretty big because it tries to move from custom Mono C APIs to native C# marshalling as much as possible, because it's now supported, it's safer, and it's more future-proof. We should be able to eventually remove the `bindings.c` layer altogether once `UnmanagedCallersOnly` and `DllImport` are handled automatically and generate Wasm imports/exports by .NET's PInvoke, which is already available in .NET alpha channel. It also provides better foundation for NativeAOT support where Mono APIs don't exist and where otherwise we'd have to maintain two very different paths for AOT and non-AOT bindings.

The performance and instruction cost of the resulting bundle is more or less the same as before, because it's still using an interpreter bundled together with .NET bytecode, but thanks to various trimming options the Wasm itself is now ~3 times smaller.

# API and ABI breaking changes

I don't think we have the right category for this, but after this change users will need to install .NET 8 to build new SpacetimeDB projects, but we'll automate the rest, e.g. installing the WASI SDK (Clang compiler & C stdlib) as well as Wasi workload for .NET itself.

I'll tentatively mark it as ABI change to draw attention to breakage even though ABI of the generated module is still compatible.

# Expected complexity level and risk

3

*This complexity rating applies not only to the complexity apparent in the diff,
but also to its interactions with existing and future code.*

*If you answered more than a 2, explain what is complex about the PR,
and what other components it interacts with in potentially concerning ways.*

### CODE REVIEW

File: crates/bindings-csharp/Runtime/bindings.c [hunk 1]
```diff
@@ -1,781 +1,231 @@
...
+       (uint32_t id, Buffer caller_identity, Buffer caller_address,
+        uint64_t timestamp, Buffer args),
+       &id, &caller_identity, &caller_address, &timestamp, &args);
 
 // Shims to avoid dependency on WASI in the generated Wasm file.
 
 #include <stdlib.h>
 #include <wasi/api.h>
 
+#pragma clang diagnostic ignored "-Wc2x-extensions"
```
Comment (kulakowski) [inline crates/bindings-csharp/Runtime/bindings.c:120]: Maybe comment what c20 feature you're using.

File: crates/bindings-csharp/Runtime/bindings.c [hunk 2]
```diff
@@ -1,781 +1,231 @@
...
-                                                   int32_t arg2,
-                                                   int64_t arg3,
-                                                   int32_t arg4) {
+// For `fd_write`, we need to at least collect and report sum of sizes.
+// If we report size 0, the caller will assume that the write failed and will
+// try again, which will result in an infinite loop.
+int32_t WASI_NAME(fd_write)(__wasi_fd_t fd, const __wasi_ciovec_t* iovs,
+                            size_t iovs_len, __wasi_size_t* retptr0) {
+  for (size_t i = 0; i < iovs_len; i++) {
+    // Note: this will produce ugly broken output, but there's not much we can
```
Comment (kulakowski) [inline crates/bindings-csharp/Runtime/bindings.c:206]: Maybe a p2 ticket tracking this?

File: crates/bindings-csharp/Runtime/Runtime.cs [hunk 3]
```diff
@@ -4,131 +4,74 @@ namespace SpacetimeDB;
...
         }
 
         public void Dispose()
         {
-            BufferIterDrop(ref handle);
+            if (!handle.Equals(RawBindings.BufferIter.INVALID))
+            {
+                RawBindings._iter_drop(handle);
+                handle = RawBindings.BufferIter.INVALID;
+                GC.SuppressFinalize(this);
```
Comment (kulakowski) [inline crates/bindings-csharp/Runtime/Runtime.cs:67]: Worth a comment I think.

File: crates/bindings-csharp/Runtime/RawBindings.cs [hunk 4]
```diff
@@ -0,0 +1,216 @@
+namespace SpacetimeDB;
+
+using System;
+// using System.Runtime.CompilerServices;
```
Comment (kulakowski) [inline crates/bindings-csharp/Runtime/RawBindings.cs:4]: Can this be deleted?

**Review by kulakowski — COMMENTED** [review]
Small code comments but looks really great!

The only thing I want is some approval from I guess @jdetter or @cloutiertyler that the download strategy is ok from a product perspective.

File: crates/bindings-csharp/Runtime/bindings.c (hunk 2 above)
Comment (RReverser) [reply crates/bindings-csharp/Runtime/bindings.c:206]: We already have one + implementation in a branch I should complete, but I've been waiting for migration to Wasmtime to happen first as it's much easier to do custom WASI on top of it.

**Comment by RReverser** [comment]
> that the download strategy is ok from a product perspective.

FWIW this is the same download strategy the previously used Wasi.Sdk package uses. I'm open to better suggestions, just pointing out that functionally it's not something new.

**Comment by kulakowski** [comment]
@RReverser Let me know if there's anything I can do to help test or land this.

**Comment by RReverser** [comment]
@jdetter wanted to do some more testing after my fixes - we ran into some issues, but they were all with unclear error messages e.g. due to .NET 8 not being installed, .NET requiring sudo on macOS to install workloads and so on.

I fixed those so in principle nothing prevents this from getting merged now, but I'd prefer to wait for @jdetter to double-check that it's looking good.

**Comment by RReverser** [comment]
The main problem is actually that I don't have macOS to test that sudo issue is gone now. @kulakowski if you want to test this, you can try to check out, do `cargo run -- build -p .\modules\spacetimedb-quickstart-cs` locally and see if error messages you get on each step make sense.

**Comment by kulakowski** [comment]
> `cargo run -- build -p .\modules\spacetimedb-quickstart-cs`

Trying now.

**Comment by RReverser** [comment]
> Trying now.

@kulakowski Any luck?

**Comment by kulakowski** [comment]
Xcode has a couple levels, the lowest is essentially just a toolchain, and the new dotnet needed a newer version of the toolchain.

**Pull request merged**